validator/src/proposal_buffer.rs

```diff
@@ -0,0 +1,340 @@
+    <TValidatorNetwork as ValidatorNetwork>::PubsubId,
+);
+
+/// The buffer holding all proposals, which have been received, but were not yet needed. Any peer can at most have
+/// one proposal in this buffer.
+///
+/// Any proposal in this buffer must have a valid signature, given the signer index within the message.
+/// On taking these messages out of the buffer (see [ProposalReceiver]) it must be maintained that
+///     1. The predecessor exists
+///     2. The proposer of the proposal calculated with the predecessor vrf seed was indeed supposed
+///         to be who was stated in the original message.
+///
+/// Any message that fails any of the above should be dropped, however depending on why it failed, additional
+/// action may be advised. If the signature was incorrect for the assumed signer, the peer who relayed the
+/// message relayed invalid data and should be banned. If the assumed proposer does not match the actual proposer,
+/// the assumed proposer may be punished (as he produced faulty data, proven by the signature and the predecessor vrf).
+/// If the predecessor is unavailable (even after requesting it from the peer who originated the proposal or relayed
+/// the proposal) they both could be banned asd they failed to produce proper data upon being asked to do so.
+pub(crate) struct ProposalBuffer<TValidatorNetwork: ValidatorNetwork + 'static>
+where
+    <TValidatorNetwork as ValidatorNetwork>::PubsubId: std::fmt::Debug + Unpin,
+{
+    /// A LinkedHashMap containing a proposal per peer_id.
+    /// A single proposal could thus be in the buffer multiple times, but a single peer cannot have
+    /// more than once proposal waiting.
+    buffer: LinkedHashMap<
+        <TValidatorNetwork::NetworkType as Network>::PeerId,
+        ProposalAndPubsubId<TValidatorNetwork>,
+    >,
+
+    /// A waker in case a proposal gets entered into a buffer while some code is
+    /// already waiting to get one out of it.
+    waker: Option<Waker>,
+}
+
+impl<TValidatorNetwork: ValidatorNetwork + 'static> ProposalBuffer<TValidatorNetwork>
+where
+    <TValidatorNetwork as ValidatorNetwork>::PubsubId: std::fmt::Debug + Unpin,
+{
+    /// Creates a new ProposalBuffer, returning the [ProposalSender] and [ProposalReceiver] that share the buffer.
+    /// Blockchain and Network are necessary to do basic verification and punishments.
+    // Ignoring clippy warning: this return type is on purpose
+    #[allow(clippy::new_ret_no_self)]
+    pub fn new(
+        blockchain: Arc<RwLock<Blockchain>>,
+        network: Arc<TValidatorNetwork>,
+    ) -> (
+        ProposalSender<TValidatorNetwork>,
+        ProposalReceiver<TValidatorNetwork>,
+    ) {
+        // Create the shared buffer
+        let buffer = Self {
+            buffer: LinkedHashMap::new(),
+            waker: None,
+        };
+        // TODO does this need to be a rwlock? Mutex? should we upgradable lock in ProposalSender::send()
```

Comment:
Should probably be a `Mutex` because we have at most on reader.

Reply:
Done